pkg/controller/update_cluster.go

```diff
@@ -45,49 +47,111 @@ import (
+// but not in the cluster.
+func (c *Controller) reconcileNamespaces(cluster *myspec.M3DBCluster) error {
+	resp, err := c.namespaceClient.List()
 	if err != nil {
-		err = fmt.Errorf("error creating namespace: %v", err)
-		c.logger.Error(err.Error())
-		c.recorder.WarningEvent(cluster, eventer.ReasonFailedCreate, err.Error())
-		return false, err
+		c.logger.Error("failed to get namespace", zap.Error(err))
+		c.recorder.WarningEvent(cluster, eventer.ReasonFailSync, err.Error())
+		return err
 	}
 
-	if !created && cluster.Status.HasInitializedNamespace() {
-		// if we didn't have to create one and the status reflects it's created, do
-		// nothing
-		return false, nil
+	if err := c.cleanupNamespaces(cluster, resp.Registry); err != nil {
+		return err
 	}
 
-	// TODO(schallert): we should use resource version + generation to not
-	// update this if not necessary
-	cluster.Status.UpdateCondition(myspec.ClusterCondition{
-		Type:           myspec.ClusterConditionNamespaceInitialized,
-		Status:         corev1.ConditionTrue,
-		LastUpdateTime: c.clock.Now().UTC().Format(time.RFC3339),
-		// TODO(schallert): probably a better reason and we should strongly type
-		// them
-		Reason:  "NamespaceCreated",
-		Message: "Created namespace",
-	})
+	if err := c.createNamespaces(cluster, resp.Registry); err != nil {
+		return err
+	}
+
+	return nil
+}
 
-	c.recorder.NormalEvent(cluster, eventer.ReasonSuccessfulCreate, "Namespace created")
+// createNamespaces will attempt to create in the cluster all namespaces which
+// are present in the spec but not the cluster.
+func (c *Controller) createNamespaces(cluster *myspec.M3DBCluster, registry *dbns.Registry) error {
+	toCreate := namespacesToCreate(registry, cluster.Spec.Namespaces)
+	for _, ns := range toCreate {
+		req, err := namespace.RequestFromSpec(ns)
+		if err != nil {
+			c.logger.Error("error forming namespace request",
+				zap.String("namespace", ns.Name),
+				zap.Error(err))
 
-	_, err = c.crdClient.OperatorV1().M3DBClusters(cluster.Namespace).Update(cluster)
-	if err != nil {
-		// TODO(schallert): event here as well?
-		err := fmt.Errorf("error updating cluster namespace status: %v", err)
-		c.logger.Error(err.Error())
-		runtime.HandleError(err)
-		return false, err
+			return fmt.Errorf("error forming request for namespace '%s': %v", ns.Name, err)
+		}
+
+		err = c.namespaceClient.Create(req)
+		if err != nil {
+			c.logger.Error("error creating namespace",
+				zap.String("namespace", ns.Name),
+				zap.Error(err))
+
+			return fmt.Errorf("error creating namespace '%s': %v", ns.Name, err)
+		}
+
+		c.recorder.NormalEvent(cluster, eventer.ReasonCreating, "created namespace "+ns.Name)
 	}
 
-	// We updated an API object so tell the queue to re-enqueue.
-	return true, nil
+	return nil
+}
+
+// cleanupNamespaces will delete any namespaces in the m3db cluster that aren't
+// in the spec.
+func (c *Controller) cleanupNamespaces(cluster *myspec.M3DBCluster, registry *dbns.Registry) error {
```

Comment:
I'm not sure `cleanup` is the most effective name here since it's just deleting namespaces and not organizing them or anything else that can be related to cleaning. Maybe `prune` is a better word?

Reply:
I like that much better! I'll update